README.md
# Tailwind CSS Schematics

[![npm version](https://badge.fury.io/js/ngx-tailwind.svg)](https://www.npmjs.com/package/ngx-tailwind)
![Schematics CI](https://github.com/notiz-dev/ngx-tailwind/workflows/Node.js%20CI/badge.svg)

Simple [Angular](https://angular.io/) schematic initializing [Tailwind CSS](https://tailwindcss.com/) in your project. [Angular v11.2](https://twitter.com/angular/status/1359736376581840896) includes native support for Tailwind CSS.

For all versions **before** Angular v11.2 a custom webpack config is added to your build process.

## Installation

Run

```bash
ng add ngx-tailwind

# or

ng add ngx-tailwind --project <MY_PROJECT>
```

Example output

```bash
The package ngx-tailwind@dev will be installed and executed.
Would you like to proceed? Yes
✔ Package successfully installed.
? Which stylesheet format are you using? SCSS [ https://sass-lang.com/documentation/syntax#scss ]
? Which @tailwindcss plugins do you want to install? typography
CREATE tailwind.config.js (236 bytes)
UPDATE package.json (1095 bytes)
UPDATE src/styles.scss (177 bytes)
✔ Packages installed successfully.
```

## Upgrade Guide

### Tailwind CSS v2 to v3

To upgrade your project from [Tailwind CSS v2 to v3](https://tailwindcss.com/docs/upgrade-guide).

```bash
npm install -D tailwindcss@latest postcss@latest autoprefixer@latest
```

Update your `tailwind.config.js` to the new [JIT Engine](https://tailwindcss.com/docs/upgrade-guide#migrating-to-the-jit-engine).

```diff
module.exports = {
- // mode: 'jit',
- purge: ['./src/**/*.{html,ts}'],
+ content: ['./src/**/*.{html,ts}'],
- darkMode: false, // or 'media' or 'class'
  theme: {
    extend: {},
  },
- variants: {
-   extend: {},
- },
  plugins: [],
};
```

Remove production build script from `package.json`, which was used for purging unused styles. Since JIT Engine generates only used styles this script is not necessary anymore.

```diff
{
  "name": "angular-workspace",
  "version": "0.0.0",
  "scripts": {
    "ng": "ng",
    "start": "ng serve",
    "build": "ng build",
    "watch": "ng build --watch --configuration development",
-   "build:prod": "cross-env NODE_ENV=production ng build --configuration production"
  },
  "private": true,
  "dependencies": {
    "@angular/animations": "~13.0.3",
    "@angular/common": "~13.0.3",
    "@angular/compiler": "~13.0.3",
    "@angular/core": "~13.0.3",
    "@angular/forms": "~13.0.3",
    "@angular/platform-browser": "~13.0.3",
    "@angular/platform-browser-dynamic": "~13.0.3",
    "@angular/router": "~13.0.3",
    "rxjs": "~7.4.0",
    "tslib": "^2.3.1",
    "zone.js": "~0.11.4"
  },
  "devDependencies": {
    "@angular-devkit/build-angular": "~13.0.4",
    "@angular/cli": "~13.0.4",
    "@angular/compiler-cli": "~13.0.3",
    "@types/node": "^12.11.1",
    "autoprefixer": "^10.4.1",
-   "cross-env": "^7.0.3",
    "postcss": "^8.4.5",
    "tailwindcss": "^3.0.8",
    "typescript": "~4.4.4"
  }
}
```

## Migrate from Tailwind CSS

To upgrade you project from [Tailwind CSS v1.x to v2.0](https://tailwindcss.com/docs/upgrading-to-v2) run the following install command

```bash
npm i -D tailwindcss@latest autoprefixer@latest postcss@latest postcss-import@latest postcss-loader@latest

# using scss
npm i -D postcss-scss@latest
```

Read the full [Upgrade Guide](https://tailwindcss.com/docs/upgrading-to-v2) to update your custom `tailwind.config.js` (e.g. [your color palette](https://tailwindcss.com/docs/upgrading-to-v2#configure-your-color-palette-explicitly)) and replace removed classes from your template (e.g. [shadow-outline and shadow-xs](https://tailwindcss.com/docs/upgrading-to-v2#replace-shadow-outline-and-shadow-xs-with-ring-utilities)).

## Additional options

You can pass additional flags to customize the schematic. For example, if you want to install a different version for **Tailwind** use `--tailwindVersion` flag:

```bash
ng add ngx-tailwind --tailwindVersion 3.0.8
```

All available flags:

| Flag                   | Description                                                    | Type            | Default                                               |
| ---------------------- | -------------------------------------------------------------- | --------------- | ----------------------------------------------------- | --- |
| `autoprefixerVersion`  | The autoprefixer version to be installed.                      | string          | `^10.4.1`                                             |
| `cssFormat`            | The file extension or preprocessor to use for style files.     | `css` \| `scss` | `css`                                                 | ‚   |
| `project`              | The project to initialize with Tailwind CSS.                   | `string`        | **First** Angular project                             |
| `postcssVersion`       | The postcss version to be installed.                           | `string`        | `^8.4.5`                                              |     |
| `tailwindVersion`      | The Tailwind version to be installed.                          | `string`        | `^3.0.8`                                              |
| `disableCrossPlatform` | Set the build:prod script to be only UNIX compatible.          | `boolean`       | `false`                                               |
| `crossEnvVersion`      | The cross-env version to be installed.                         | `string`        | `^7.0.3`                                              |
| `tailwindPlugins`      | @tailwindcss plugins installed and added to tailwind.config.js | `string[]`      | [`aspect-ratio`, `forms`, `line-clamp`, `typography`] |

Advanced usage

```bash
ng add ngx-tailwind --cssFormat scss --tailwindVersion 3.0.8 --postcssVersion 8.4.5
```

Want to integrate Tailwind CSS with Angular 11.1 or lower? No problem:

```bash
ng add ngx-tailwind@2.3.0
```

By default, `cross-env` is added to the `build:prod` script to be able to set `NODE_ENV=prod` cross-platform.
If you want to override the default behavior, you can set the flag `--disableCrossPlatform`:

```bash
ng add ngx-tailwind --disableCrossPlatform
```

## Developing

Install `@angular-devkit/schematics-cli` to be able to use `schematics` command

```bash
npm i -g @angular-devkit/schematics-cli
```

Now build the schematics and run the schematic.

```bash
npm run build
# or
npm run build:watch

# dry-run in angular-workspace
npm run start:dev

# execute schematics in angular-workspace
npm run start
```


CHANGELOG.md
# [4.0.0](https://github.com/notiz-dev/ngx-tailwind/compare/v3.0.0...v4.0.0) (2022-07-12)

### Features

* **tailwind**: add tailwind types to `tailwind.config.js``
* **angular**: update to angular 14

# [3.0.0](https://github.com/notiz-dev/ngx-tailwind/compare/v2.3.0...v3.0.0) (2022-01-02)

### Features

* **tailwind**: update to tailwind 3
* **angular**: remove support for angular below 11.2
* **angular**: remove unnecessary angular production build script

# [2.3.0](https://github.com/notiz-dev/ngx-tailwind/compare/v2.2.2...v2.3.0) (2021-12-04)

### Features

* **angular**: update to angular 13
* **tailwind**: update version to latest

## [2.2.2](https://github.com/notiz-dev/ngx-tailwind/compare/v2.2.1...v2.2.2) (2021-09-30)


### Features

* **schematics:** update default versions ([38006ce](https://github.com/notiz-dev/ngx-tailwind/commit/38006ce06123b95c312878822e9fd92093ced4b0))



## [2.2.1](https://github.com/notiz-dev/ngx-tailwind/compare/v2.2.0...v2.2.1) (2021-05-18)


### Bug Fixes

* **css:** tailwind directives name ([20896a4](https://github.com/notiz-dev/ngx-tailwind/commit/20896a478ac79aacb860378a1461d9d16bcb4375))



# [2.2.0](https://github.com/notiz-dev/ngx-tailwind/compare/v2.1.0...v2.2.0) (2021-05-17)


### Features

* **cli:** check @angular/cli version ([d904226](https://github.com/notiz-dev/ngx-tailwind/commit/d904226c52b68e8fbc2b375c545f231d8c391812))
* **tailwind:** add jit mode as comment to tailwind config ([388f1cf](https://github.com/notiz-dev/ngx-tailwind/commit/388f1cf98c2550e3a1dca338c40e9f9766d20ec0))



# [2.1.0](https://github.com/notiz-dev/ngx-tailwind/compare/v2.0.0...v2.1.0) (2021-02-15)


### Features

* **angular-cli:** prompt if using angular cli 11.2 or above ([fbf4637](https://github.com/notiz-dev/ngx-tailwind/commit/fbf463783c62ed283660ad2f5ea4ef862ed97263))
* **schematics:** new setup for angular 11.2 and above ([47225ba](https://github.com/notiz-dev/ngx-tailwind/commit/47225ba5e2d0aed39000655406cf35205bd9b60c))



# [2.0.0](https://github.com/notiz-dev/ngx-tailwind/compare/v1.3.0...v2.0.0) (2021-01-21)


### Bug Fixes

* **workspace:** new workspace utils import for angular 11 ([5926df3](https://github.com/notiz-dev/ngx-tailwind/commit/5926df3c93e294b179105ec913b874dd7e278736))


### Features

* **deps:** update to latest deps ([681b5eb](https://github.com/notiz-dev/ngx-tailwind/commit/681b5ebd3bccd847c21b4008325a17a8175ea359))



# [1.3.0](https://github.com/notiz-dev/ngx-tailwind/compare/v1.2.0...v1.3.0) (2021-01-19)


### Features

* **plugins:** select and install [@tailwindcss](https://github.com/tailwindcss) core plugins ([97716c8](https://github.com/notiz-dev/ngx-tailwind/commit/97716c8870cfb9ac7e3a1948f2c9725d113b6a9e))



# [1.2.0](https://github.com/notiz-dev/ngx-tailwind/compare/v1.1.0...v1.2.0) (2021-01-13)


### Bug Fixes

* **cross-env:** modify the flag installCrossPlatform to disableCrossPlatform ([af3d717](https://github.com/notiz-dev/ngx-tailwind/commit/af3d71727fa65e2642e3e8991a8a985cd3366832))


### Features

* **cross-env:** add cross-env to build:prod script ([2be0001](https://github.com/notiz-dev/ngx-tailwind/commit/2be0001e0737509aa57c41510de4fc972b8209c2))
* **deps:** update dependencies ([651fda8](https://github.com/notiz-dev/ngx-tailwind/commit/651fda8df9eb1276a3d2980420632640b1db825f))



# [1.1.0](https://github.com/notiz-dev/ngx-tailwind/compare/v1.0.0...v1.1.0) (2020-12-11)


### Features

* **tailwind:** add simple purge option ([a6e309b](https://github.com/notiz-dev/ngx-tailwind/commit/a6e309ba07fb5fd046341773cb05b0e4a933106b))



# [1.0.0](https://github.com/notiz-dev/ngx-tailwind/compare/v0.5.0...v1.0.0) (2020-11-18)


### Features

* **dependency:** install autoprefixer@^10.0.2 ([210fa34](https://github.com/notiz-dev/ngx-tailwind/commit/210fa34df0ccd44f6a9ebe63e34d3ab74a85adec))
* **dependency:** update default dependencies ([4395758](https://github.com/notiz-dev/ngx-tailwind/commit/4395758b3d7cd34a200d328686a9a5ef19839bfa))
* **npm-script:** add purge script for tailwind ([f13c66a](https://github.com/notiz-dev/ngx-tailwind/commit/f13c66a52c2f3331b7ed43e5e43b4564bf3898a3))
* **postcss:** install postcss ([dd4042b](https://github.com/notiz-dev/ngx-tailwind/commit/dd4042bc517285c581cab55b9ce7618c02ef9fef)), closes [#8](https://github.com/notiz-dev/ngx-tailwind/issues/8)



## [0.5.1](https://github.com/notiz-dev/ngx-tailwind/compare/v0.5.0...v0.5.1) (2020-10-23)


### Features

* **npm-script:** add purge script for tailwind ([f13c66a](https://github.com/notiz-dev/ngx-tailwind/commit/f13c66a52c2f3331b7ed43e5e43b4564bf3898a3))



# [0.5.0](https://github.com/notiz-dev/ngx-tailwind/compare/v0.2.0...v0.5.0) (2020-10-23)


### Features

* **ng-add:** save lib as devDeps ([120f553](https://github.com/notiz-dev/ngx-tailwind/commit/120f5538d38466d61990505fec0e80b26a17e62c))
* **ngx-tailwind:** rename lib ([70adecf](https://github.com/notiz-dev/ngx-tailwind/commit/70adecf858a23e2f97de5613ffa9e64130f1026a))
* **schematics:** add additional version options ([ef07ab9](https://github.com/notiz-dev/ngx-tailwind/commit/ef07ab90c9319847064a7e0942f11bfa6965f764))
* **stylesheet:** select css stylesheet format ([f8c026a](https://github.com/notiz-dev/ngx-tailwind/commit/f8c026a3cdfec7a3a56fd26a97abfc45f937983d))
* **tailwind:** add tailwind-init task ([3aebe1a](https://github.com/notiz-dev/ngx-tailwind/commit/3aebe1a58987230bcc046dcdbb4ea05f80a2795f))
* **webpack:** update postcss options ([d7b7389](https://github.com/notiz-dev/ngx-tailwind/commit/d7b7389be3ff87d75015d87f699c47f0f05f2e57))





src/ng-add/index.ts
import { strings } from '@angular-devkit/core';
import {
  Rule,
  SchematicContext,
  Tree,
  chain,
  SchematicsException,
  url,
  apply,
  mergeWith,
  template,
} from '@angular-devkit/schematics';
import { Schema } from './schema';
import {
  getProjectFromWorkspace,
  getProjectStyleFile,
} from '@angular/cdk/schematics';
import { NodePackageInstallTask } from '@angular-devkit/schematics/tasks';
import {
  addPackageJsonDependency,
  NodeDependencyType,
} from '@schematics/angular/utility/dependencies';
import { getWorkspace } from '@schematics/angular/utility/workspace';
import { InsertChange } from '@schematics/angular/utility/change';
import { WorkspaceDefinition } from '@angular-devkit/core/src/workspace';
import { hasTailwindSupport } from '../util/has-tailwind-support';

export function ngAdd(_options: Schema): Rule {
  return async (host: Tree) => {
    const workspace = await getWorkspace(host);

    if (!_options.project) {
      _options.project = workspace.projects.keys().next().value;
    }

    if (!_options.project) {
      throw new SchematicsException(
        `Project ${_options.project} is not defined in this workspace.`,
      );
    }

    const tailwindPlugins = _options.tailwindPlugins || [];
    const tailwindPluginDependencies = tailwindPlugins?.map(
      (plugin) => `@tailwindcss/${plugin}`,
    );
    const requireTailwindPlugins = tailwindPluginDependencies.map(
      (plugin) => `require('${plugin}')\n`,
    );

    if (hasTailwindSupport(host)) {
      return chain([
        addDependenciesWithTailwindSupport(_options),
        addTailwindPlugins(tailwindPluginDependencies),
        updateStyles(_options, workspace),
        generateTailwindConfig(_options, requireTailwindPlugins),
        install(),
      ]);
    } else {
      throw new SchematicsException(
        'Current version only supports Angular 11.2 or higher. Use previous version for Angular 11.1 or lower: ng add ngx-tailwind@2.3.0',
      );
    }
  };
}

function addDependenciesWithTailwindSupport(_options: Schema): Rule {
  return (host: Tree) => {
    addPackageJsonDependency(host, {
      type: NodeDependencyType.Dev,
      name: 'tailwindcss',
      version: _options.tailwindVersion,
    });

    addPackageJsonDependency(host, {
      type: NodeDependencyType.Dev,
      name: 'autoprefixer',
      version: _options.autoprefixerVersion,
    });

    addPackageJsonDependency(host, {
      type: NodeDependencyType.Dev,
      name: 'postcss',
      version: _options.postcssVersion,
    });
  };
}

function addTailwindPlugins(tailwindPlugins: string[]): Rule {
  return (tree: Tree) => {
    tailwindPlugins.forEach((plugin) =>
      addPackageJsonDependency(tree, {
        type: NodeDependencyType.Dev,
        name: plugin,
        version: 'latest',
      }),
    );
  };
}

function updateStyles(options: Schema, workspace: WorkspaceDefinition): Rule {
  return (tree: Tree, context: SchematicContext) => {
    const project = getProjectFromWorkspace(workspace, options.project);
    const stylePath = getProjectStyleFile(project, options.cssFormat);

    if (!stylePath) {
      context.logger.error(
        `Cannot update project styles file: Style path not found`,
      );
      return tree;
    }

    const insertion = new InsertChange(
      stylePath!,
      0,
      options.cssFormat === 'css'
        ? getTailwindDirectives()
        : getTailwindImports(),
    );
    const recorder = tree.beginUpdate(stylePath!);
    recorder.insertLeft(0, insertion.toAdd);
    tree.commitUpdate(recorder);

    return tree;
  };
}

/**
 * Used for css stylesheets
 */
function getTailwindDirectives(): string {
  return `@tailwind base;\n
@tailwind components;\n
@tailwind utilities;\n`;
}

/**
 * Used for scss stylesheets
 */
function getTailwindImports(): string {
  return `@import 'tailwindcss/base';\n
@import 'tailwindcss/components';\n
@import 'tailwindcss/utilities';\n`;
}

/**
 * Generate webpack and tailwind config.
 *
 * @param options
 */
function generateTailwindConfig(
  options: Schema,
  requireTailwindPlugins: string[],
): Rule {
  return async (_host: Tree) => {
    const sourceTemplates = url(`./templates/tailwind`);
    const sourceParametrizedTemplates = apply(sourceTemplates, [
      template({
        ...options,
        requireTailwindPlugins: requireTailwindPlugins,
        ...strings,
      }),
    ]);
    return mergeWith(sourceParametrizedTemplates);
  };
}

function install(): Rule {
  return async (_host: Tree, context: SchematicContext) => {
    // Install the dependency
    context.addTask(new NodePackageInstallTask());
  };
}
